### ISSUE TITLE
Verify throws exception when non-transient registrations depend on RegisterWithContext registrations

### ISSUE THREAD

**Issue by dotnetjunkie** [simpleinjector/SimpleInjector#103]
[Question from [azzlack](https://github.com/azzlack) asked originally posted on #102]

The RegisterWithContext extension does not work with v3 when using WebAPI and classes have `WebApiRequestLifestyle`, because the extensions uses the `Transient` lifestyle. 

```csharp
container.RegisterWithContext<ILog>(
    dependencyContext =>
    {
        if (dependencyContext.ImplementationType != null)
        {
            return LogManager.GetLogger(dependencyContext.ImplementationType);
        }

        return LogManager.GetCurrentClassLogger();
    });
```

Is there any way to solve this?

I've tried to disable warnings by writing

```csharp
var registration = container.GetRegistration(typeof(ILog)).Registration;
registration.SuppressDiagnosticWarning(DiagnosticType.LifestyleMismatch,
     "The logger must be new every time");
```

but I still get the warnings/errors

**Comment by dotnetjunkie** [comment]
The `RegisterWithContext` extension method from the old 2.x documentation needs to make registrations transient, because the registration can depend on runtime conditions that the user can make inside the registered delegate. Making the lifestyle anything else than transient, can lead to very weird behavior. But this will result in being forced to make your complete object graph transient.

Because of this the use of `RegisterWithContext` is discouraged in Simple Injector 3 (the example is removed from the documentation). In Simple Injector 3 you are advised to use the new `RegisterConditional` method overloads. The `RegisterConditional` method overloads are limited to do auto-wiring; you can't register a delegate. This is deliberate, because we want to prevent users making runtime decisions that are unverifiable by Simple Injector. Instead you will have to register a type, which can be done as follows:

``` c#
container.RegisterConditional(typeof(ILog),
    c => typeof(Logger<>).MakeGenericType(c.Consumer.ImplementationType),
    Lifestyle.Singleton,
    c => true);
```
Note: This example is copied from [the documentation](https://simpleinjector.readthedocs.org/en/latest/advanced.html#context-based-injection).

To make this work, you will have to create the `Logger<T>` type, but this is trivial:

``` c#
public sealed class Logger<T> : ILog
{
    private readonly ILog logger = LogManager.GetLogger(typeof(T));

    // ILog members
    public void Log(string message) {
        this.logger.Log(message);
    }
}
```

**Comment by azzlack** [comment]
The documentation says to use `Lifestyle.Transient` (and that's what I've tried) but will check out how it handles with `Lifestyle.Singleton` :-)

**Comment by azzlack** [comment]
Ok. So Singleton gets rid of the error, but now `T` resolves to the same type every time.  
But what do you mean by "you can't register a delegate"? 

I'm now using the following code:

```csharp
container.RegisterConditional(
    typeof(ILog),
    c =>
    {
        if (c.Consumer != null)
        {
            return typeof(MyLogger<>).MakeGenericType(c.Consumer.ImplementationType);
        }

        return typeof(MyLogger<>).MakeGenericType(typeof(ILog));
    },
    Lifestyle.Singleton,
    c => true);
```

The reason for the null check is that when running `Startup.cs` in an OWIN app the Consumer is null

**Comment by dotnetjunkie** [comment]
What I meant with "you can't register a delegate" is, that you can't register a delegate that will be called when the object graph is resolved. Of course that particular overload accepts two delegates, but these are not the 'instanceCreator' delegate. The last delegate is the predicate that determines whether the registration holds in a particular context. The first delegate is the 'type factory'. Simple Injector will call that delegate once per consumer during the compilation process to determine the type to resolve. After that, this information is burned in the compiled object graph and the delegate won't be called again.

The `Consumer` property will be `null` in the registration is resolved directly from the container (and therefore doesn't have a consumer). In other words, if you call `container.GetInstance<ILog>()`, `c.Consumer` will be null. Usually however, you will never resolve a logger directly; it will always be injected into something else. In that case you don't have to check `Consumer != null`.

**Comment by azzlack** [comment]
Ok. So if I remove the null check it should work?

**Comment by dotnetjunkie** [comment]
> Ok. So if I remove the null check it should work?

As long as you don't resolve `ILog` directly: yes.

> But it still resolves to the same T every time.

That's a bug. I've been able to reproduce this. That's very unfortunate. We will investigate and fix a.s.a.p.

**Comment by dotnetjunkie** [comment]
Added a new work item for this #104.

**Comment by azzlack** [comment]
Thanks! I'll be able to live with the wrong types for now, but really looking forward to the fix :-)

**Comment by azzlack** [comment]
That reminds me, the `SuppressDiagnosticWarning` is not working as well. Or am I using it wrong?

**Comment by dotnetjunkie** [comment]
Can you elaborate on the issue with `SuppressDiagnosticWarning`?

**Comment by azzlack** [comment]
Given the following (this is in a WebAPI project):

```csharp
var container = new Container();
container.Register<ILog, Log>(Lifestyle.Transient);
container.RegisterWebApiRequest<ILog, Log>();

var registration = container.GetRegistration(typeof(ILog)).Registration;
registration.SuppressDiagnosticWarning(DiagnosticType.LifestyleMismatch, "The logger must be new every time");

container.Verify();
```

```csharp
public class UserFactory
{
    private readonly ILog log;

    public UserFactory(ILog log)
    {
        this.log = log;
    }
}
```

This gives the following error:

```
The configuration is invalid. The following diagnostic warnings were reported:
-[Lifestyle Mismatch] UserFactory (Web API Request) depends on ILog implemented by Log (Transient).
```

**Comment by dotnetjunkie** [comment]
The suppression needs to be made on the `UserFactory` like this:

``` c#
container.GetRegistration(typeof(UserFactory)).Registration
```

But this might actually be a design flaw. I will have to discuss this with my team. To be continued.

**Comment by azzlack** [comment]
Ah. Yeah that seems like a hassle if you have to suppress for every place `ILog` is used. In my case this would be like 30-40 classes.

Looking forward to what your discussion leads to :-)

**Comment by qujck** [comment]
##### In the red corner we have the implemented behaviour:

```csharp
var registration = container.GetRegistration(typeof(UserFactory)).Registration;
registration.SuppressDiagnosticWarning(DiagnosticType.LifestyleMismatch, 
    "The logger must be new every time");
```

> it is safe for *UserFactory* to hold *ILog* as a captive dependency

##### In the blue corner we have the desired behaviour:

```csharp
var registration = container.GetRegistration(typeof(ILog)).Registration;
registration.SuppressDiagnosticWarning(DiagnosticType.LifestyleMismatch, 
    "The logger must be new every time");
```

> *ILog* can always be a captive dependency

It seems to me that both these options are entirely plausible; the question I feel we should ask is can `SuppressDiagnosticWarning` be enhanced to offer this finer level of control? (Consider marking the current method as `[Obsolete]`.)

**Comment by qujck** [comment]
Upon reflection the red corner is ambiguous because it states that `UserFactory` can reference objects with as lifestyle that may cause unintended side effects.

**Comment by dotnetjunkie** [comment]
I think the blue behavior is most intuitive to users.

Do note that the reason that the [Lifestyle Mismatch documentation](https://simpleinjector.readthedocs.org/en/latest/LifestyleMismatches.html) doesn't include an example of the suppression is because this warning should not be suppressed. Or as the docs put it:

> Do not ignore these warnings. False positives for this warning are rare and even when they occur, the registration or the application design can always be changed in a way that the warning disappears.

Exception to the rule might be the use of the `RegisterWithContext` method, but this method is obsolete, and when #102 is implemented, there is no reason left to use `RegisterWithContext`.

**Comment by dotnetjunkie** [comment]
The two bugs #104 and #108 that were reported in this thread have been fixed in [v3.0.3](https://github.com/simpleinjector/SimpleInjector/releases/tag/v3.0.3). This version can be downloaded from NuGet.

Thanks for reporting this @azzlack.

**Comment by azzlack** [comment]
The new version works perfectly. Thanks! @dotnetjunkie @qujck
